src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala

```diff
@@ -34,76 +32,160 @@ trait WorkerMain[S] {
           stdout.synchronized {
-            WorkerProtocol.WorkResponse.newBuilder
+            val builder = WorkerProtocol.WorkResponse.newBuilder
+            builder
               .setRequestId(requestId)
-              .setOutput(outStream.toString)
-              .setExitCode(code)
-              .build
+              .setWasCancelled(wasCancelled)
+
+            maybeOutStream.foreach { outStream =>
+              builder.setOutput(outStream.toString)
+            }
+
+            maybeExitCode.foreach { exitCode =>
+              builder.setExitCode(exitCode)
+            }
+
+            builder
+              .build()
               .writeDelimitedTo(stdout)
           }
+
+          activeRequests.remove(requestId)
         }
 
+        /**
+         * Think of this function as a single threaded event loop that gets work from Bazle and farms out work to the
+         * multi threaded execution context defined above.
+         */
         @tailrec
         def process(ctx: S): Unit = {
           val request = WorkerProtocol.WorkRequest.parseDelimitedFrom(stdin)
           if (request == null) {
             return
           }
-          val args = request.getArgumentsList.toArray(Array.empty[String])
-          val sandboxDir = Path.of(request.getSandboxDir())
-
-          // We go through this hullabaloo with output streams being defined out here, so we can
-          // close them after the async work in the Future is all done.
-          // If we do something synchronous with Using, then there's a race condition where the
-          // streams can get closed before the Future is completed.
-          var outStream: ByteArrayOutputStream = null
-          var out: PrintStream = null
 
           val requestId = request.getRequestId()
-          System.out.println(s"WorkRequest $requestId received with args: ${request.getArgumentsList}")
-
-          val f: Future[Int] = Future {
-            outStream = new ByteArrayOutputStream
-            out = new PrintStream(outStream)
-            try {
-              work(ctx, args, out, sandboxDir)
-              0
-            } catch {
-              case e @ AnnexWorkerError(code, _, _) =>
-                e.print(out)
-                code
+
+          // If this is a cancel request, we need to cancel a previously sent WorkRequest
+          // Arguments and inputs fields on cancel requests "must be empty and should be ignored"
+          if (request.getCancel()) {
+            System.err.println(s"Cancellation WorkRequest received for request id: $requestId")
+
+            // From the Bazel doc: "The server may send cancel requests for requests that the worker
+            // has already responded to, in which case the cancel request must be ignored."
+            Option(activeRequests.get(requestId)).foreach { activeRequest =>
+              // Cancel will wait for the thread to complete or be interrupted, so we do it in a future
```

Comment:
Does this defeat the purpose of waiting for the thread to complete (which was described in the comment above `FutureTaskWaitOnCancel`), or am I misunderstanding how this works?

Reply:
We haven't responded to Bazel that the task is cancelled, so I believe we're still good. Once we respond that things are cancelled, then I believe Bazel knows this worker has a slot to send work to.